src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -1431,3 +1434,44 @@ def rollback_impl(self, args, id):
             )
         except subprocess.CalledProcessError as e:
             raise CallError(f'Failed to rollback snapshot: {e.stderr.strip()}')
+
+    @accepts(
+        Str('id'),
+        Dict(
+            'options',
+            Bool('recursive', default=False),
+        ),
+    )
+    def hold(self, id, options):
+        """
+        Holds snapshot `id`.
+
+        `truenas` tag will be added to the snapshot's tag namespace.
+        """
+        try:
+            with libzfs.ZFS() as zfs:
+                snapshot = zfs.get_snapshot(id)
+                snapshot.hold('truenas', options['recursive'])
+        except libzfs.ZFSException as err:
```

Comment:
Since we're hardcoding tags, do we need special behavior for failure case where hold is already set?

Reply:
A special case that does what, no-op?

Reply:
It was merely a question. I was looking through in what situations it fails in `dsl_dataset_user_hold_check()`. If you don't think anything is required that's fine.